python/tests/test_topology.py

```diff
@@ -7215,18 +7215,54 @@ def test_reference_sequence(self):
 class TestConcatenate:
     def test_simple(self):
         ts1 = tskit.Tree.generate_comb(5, span=2).tree_sequence
+        ts1 = msprime.sim_mutations(ts1, rate=1, random_seed=1)
         ts2 = tskit.Tree.generate_balanced(5, arity=3, span=3).tree_sequence
+        ts2 = msprime.sim_mutations(ts2, rate=1, random_seed=2)
         assert ts1.num_samples == ts2.num_samples
         assert ts1.num_nodes != ts2.num_nodes
         joint_ts = ts1.concatenate(ts2)
         assert joint_ts.num_nodes == ts1.num_nodes + ts2.num_nodes - 5
         assert joint_ts.sequence_length == ts1.sequence_length + ts2.sequence_length
         assert joint_ts.num_samples == ts1.num_samples
+        assert joint_ts.num_sites == ts1.num_sites + ts2.num_sites
+        assert joint_ts.num_mutations == ts1.num_mutations + ts2.num_mutations
         ts3 = joint_ts.delete_intervals([[2, 5]]).rtrim()
         # Have to simplify here, to remove the redundant nodes
-        assert ts3.equals(ts1.simplify(), ignore_provenance=True)
+        ts3.tables.assert_equals(ts1.tables, ignore_provenance=True)
         ts4 = joint_ts.delete_intervals([[0, 2]]).ltrim()
-        assert ts4.equals(ts2.simplify(), ignore_provenance=True)
+        ts4.tables.assert_equals(ts2.tables, ignore_provenance=True)
+
+    def test_metadata(self, ts_fixture):
+        tables = ts_fixture.dump_tables()
+        tables.reference_sequence.clear()
+        tables.migrations.clear()
+        ts = tables.tree_sequence()
+        num_sites = ts.num_sites
+        assert num_sites > 0
+        joint_ts = ts.concatenate(ts)
+        for s1, s2 in zip(range(num_sites), range(num_sites, num_sites * 2)):
+            site1 = joint_ts.site(s1)
+            site2 = joint_ts.site(s2)
+            assert site1.metadata == site2.metadata
+            assert site1.ancestral_state == site2.ancestral_state
+            assert len(site1.mutations) == len(site2.mutations)
+            for m1, m2 in zip(site1.mutations, site2.mutations):
+                assert m1.metadata == m2.metadata
+                assert m1.derived_state == m2.derived_state
+                assert m1.time == m2.time
+        ns_nodes = np.where(ts.tables.nodes.flags & tskit.NODE_IS_SAMPLE == 0)[0]
+        assert len(ns_nodes) > 0
+        for u1, u2 in zip(ns_nodes[: len(ns_nodes)], ns_nodes[len(ns_nodes) :]):
```

Comment:
I think the second argument is an empty slice, so this loop won't run. It is wise to always use the `strict` argument to detect these kind of things.